readme.md
# Quantogram - Approximate Quantile calculation using Histograms

## A library for Estimating Online Quantiles of Streams

`Quantogram` accepts a stream of floating point numbers (`f64`) with optional weights and estimates the desired **quantile**, such as the **median**, as well as providing other common measures like **count**, **min**, **max**, **mean** and **mode**. The caller may tradeoff storage and accuracy by choosing the desired absolute relative error from this range:

  - lowest accuracy = 17.2%
  - default accuracy = 1.0%
  - highest accuracy = 0.034%

## Installation

Add this to your Cargo.toml:
```
[dependencies]
quantogram = "0.4"
```
then you are good to go.

The library requires Rust's 2021 edition, but only so that it can run benchmarks against the `zw-fast-quantile` library, which requires that version. You could build from source and remove the benchmarks related to ZW if you need the 2018 edition.

## Release Notes

See the change log for changes: 
[Change log](./changelog.md)

## Usage

The default configuration is adequate for most uses and promises a 1% worst-case error rate.

**Default Quantogram with Unweighted Samples**

This example creates a default `Quantogram`, adds some samples, and computes the count, min, max, mean, median, and third quartile (75% quantile). It also removes some samples and repeats the median.

```rust
    use quantogram::Quantogram;
    let mut q = Quantogram::new();
    q.add(10.0);
    q.add(40.0);
    q.add(20.0);
    q.add(30.0);
    q.add(50.0);

    assert_eq!(q.count(), 5);
    assert_eq!(q.min().unwrap(), 10.0);
    assert_eq!(q.max().unwrap(), 50.0);
    assert_eq!(q.mean().unwrap(), 30.0);
    assert_eq!(q.median().unwrap(), 30.0);
    assert_eq!(q.quantile(0.75).unwrap(), 40.0);

    q.remove(10.0);
    q.remove(20.0);
    assert_eq!(q.median().unwrap(), 40.0);
```

**Use QuantogramBuilder to set accuracy**

This example creates a `Quantogram` using a `QuantogramBuilder`, sets its accuracy to 0.5%, adds a list of samples in bulk, and computes the median.

```rust
    use quantogram::{QuantogramBuilder,Quantogram};
    let mut q = QuantogramBuilder::new()
                .with_error(0.005)
                .build();
    let data = vec![10.0,40.0,20.0,30.0,50.0];
    q.add_unweighted_samples(data.iter());
    assert_eq!(q.median().unwrap(), 30.0);
```

**Working with Weighted Samples**

This example begins by adding unweighted samples (which are assigned a weight of one) then adds in a heavily weighted zero sample.

```rust
    use quantogram::{Quantogram};
    let mut q = Quantogram::new();
    let data = vec![1.0,2.0,3.0,4.0,5.0,95.0,96.0,97.0,98.0,99.0];
    q.add_unweighted_samples(data.iter());
    q.add_weighted(0.0, 6.0);
    assert_eq!(q.median().unwrap(), 2.0); 
```

**Dealing with Gaps in data**

If there is a large gap in the data at the point where the desired quantile falls, it is proper to average the value below and above the gap to get the quantile. The `median` and `quantile` methods do not attempt this, but `fussy_quantile` does. For example, the following sequence has an even number of elements, so the median is supposed to be the average of the two middle values, 5 and 95, yielding 50, a number not even close to any of the values in the set:
```
 Samples = { 1,2,3,4,5,95,96,97,98,99 }
```

```rust
    use quantogram::{Quantogram};
    let mut q = Quantogram::new();
    let data = vec![1.0,2.0,3.0,4.0,5.0,95.0,96.0,97.0,98.0,99.0];
    q.add_unweighted_samples(data.iter());
    assert_eq!(q.fussy_quantile(0.5, 2.0).unwrap(), 50.0); 
```

## Configurable Memory Usage

The upper limit on storage required for the highest accuracy is 500x that of the lowest accuracy and 29x that of the default accuracy. One can choose any value for error rate that falls in the range above.

The storage is affected by how many histogram bins each power of two range of values is divided into. Few bins gives a coarse result while many bins gives a fine result. Storage is linearly proportional to the number of bins chosen. 

![Error formula](./error-formula.gif "Error Formula")

The growth factor indicates that each bin is wider than the previous bin in an exponential way.

![Growth formula](./growth-formula.gif "Growth Formula")

For example, if `bins` is 35 then `growth` is 1.02 and `error` is 1%.

| Bins | Error Rate |
| :--- | ---------: |
|    2 |    17.2 %  |
|    3 |    11.5 %  | 
|    4 |     8.6 %  | 
|    5 |     6.9 %  | 
|    7 |     4.9 %  | 
|   12 |     2.9 %  | 
|   18 |     1.9 %  | 
|   35 |     1.0 %  | 
|   70 |     0.5 %  |
|  139 |     0.25%  |
|  347 |     0.10%  |
|  694 |     0.05%  |
|  867 |     0.04%  |
| 1000 |     0.0347%|

To halve the error rate you need to double the number of bins (and double the storage).

The maximum storage requirement is approximately this (setting aside a few configuration properties in the structure and overhead imposed by the Skip Dictionary):

```
storage = (2 x 8-byte floats) x (2 x 254 x (bins + 1) + 7)
        ≈ 8 kb x (bins + 1)
```

The 2 x 254 has 2 for positive and negative values and 254 for the number of powers of two between the smallest and largest magnitude numbers supported by 64-bit floats. The bins + 1 is because there are two histograms, a coarse histogram that just collects one count for the nearest power of two and a fine histogram that divides each power of two into `bins` separate counts.

However, if samples never go above a million and are only recorded to three decimal places, the 254 in the formula above could be reduced to 31.

The range of maximum storage required goes from 25 KB for 2 bins to 8.2 MB for 1000 bins. For the default of 35 bins at 1% accuracy, the storage max is about 300 KB. However, if only positive integers from one to a million were added, this would drop to 17 KB. Drop the accuracy to 5% and the memory required drops to 3.5 KB. Thus you have much flexibility in tuning the structure to achieve your goals. (Fewer bins and lower accuracy also leads to faster performance.)

## Features

**Any quantile**. Some algorithms require you to choose the target quantile before you start collecting the data. This structure permits the querying of any quantile and does not require you to configure it to target any particular range of phi.

**Extreme Quantiles**. Some algorithms lose accuracy when dealing with extreme quantiles (values near zero or one) and may compensate by increasing storage. `Quantogram`'s accuracy guarantee applies to all values for phi, the desired quantile. Also, its storage requirements do not change to accommodate extreme quantiles.

**Weighted Samples**. Samples may be uniformly weighted (weight = 1) or be assigned separate weights, so that weighted medians and weighted quantiles may be computed.

**Remove Samples**. Samples may be removed and the quantiles will be updated to reflect this. 

*Warning*: If the minimum or maximum value is removed, subsequent requests for the minimum or maximum (until a new extreme sample is added) may return an incorrect value. 

**Descriptive Statistics**. `Quantogram` provides these basic statistics: 

 - **min**
 - **max**
 - **mean**
 - **median**
 - **variance**
 - **standard deviation**
 - **quantile**
 - **q1, q3** (1st and 3rd quartiles)
 - **mode**
 - **hsm** (half-sample mode)
 - **count**

*Notes*: 

 1. The standard deviation (stddev) calculates the population standard deviation, not the sample standard deviation.
 2. If only integer values are collected, the `mode` will be rounded to the nearest integer.
 3. If integer values in the range [-63,63] are used, a good value for mode will result. 
 4. If non-integer values are collected or integers outside this range, then the effect of non-uniform histogram bin widths will distort the `mode` and the anwswer will not be what you expect. Bins vary in size exponentially, so the `mode` will act like the mode of the log of the samples. This will skew the `mode` toward larger values, as larger numbers are part of larger bins.
 5. Most rules of thumb related to the use of histograms to estimate the **mode** (like **Sturges' Rule** and **Scott's Rule**) use bin counts that are much lower than what is used by `Quantogram`. It might be better to rebin by consolidating multiple adjacent bins in order to compute the `mode`. 
 6. As an alternative to the mode, try the **half-sample mode**, `Quantogram::hsm`. This applies a generalized form of the **Robertson-Cryer (1974)** *half-sample mode estimator* to histogrammed data instead of the raw samples. It has been generalized from the algorithm by **David Bickel** to apply to weighted samples. The *half-sample mode* is good at ignoring outliers and contamination.
 7. The **half-sample mode** (`hsm`) as implemented is worse than quadratic in performance (possibly `O(N² Log N)`). To compensate for this, a cache has been added that often recalculates the `hsm` over a subset of the data in the neighborhood of the previously calculated mode. 
    - If new values are added that are within this neighborhood, partial recalc will be done. If the Quantogram uses buckets with growth of 1.02 (for 1% quantization error) then the default neighborhood of 70 bins covers the range `[1/2 x mode, 2 x mode]`. If a distribution has a well-defined peak, this core region should often include the newly added sample and permit the optimization.
    - If new values are outside the neighborhood, it is unlikely that they will move the mode. Permit reuse of previously computed mode without any recalc for a dozen samples in a row before forcing a full recalc. 
    - If the number of buckets grows too much or the change in total weight is too much, a full recalc will be done.
    - If there are fewer than 64 bins, perform a full recalc. This enables a proper and accurate mode to arise from the data before the caching kicks in.

**Measures of Dispersion**

These additional statistics are implemented which are measures of dispersion:

 - **range** (spread between min and max)
 - **iqr** (interquartile range: Q3 - Q1)        
 - **quartile_deviation** (semi-interquartile range, half of iqr))
 - **coeff_of_range** ((max - min) / (max + min))
 - **coeff_of_quartile_dev** ((Q3-Q1)/(Q3+Q1))
 - **coeff_of_stddev** (standard deviation divided by mean)
 - **coeff_of_variation**

**Inverse Quantile**. Lookup the quantile at which a given value falls using the `quantile_at` function.

**Exceptional Values**. Infinities and NANs are counted separately and do not factor into the quantile calculation. The proportion of such values may be queried.

**Overflow and Underflow**. If infinitesmal values are not ineresting, the underflow value may be changed to treat values smaller than a threshold as zeroes. Likewise, an upper threshold may be set and all values larger in magnitude are considered infinities (positive or negative).

By compressing the dynamic range of recorded values, less storage is required and performance is enhanced.

**Special handling of Integers**. So long as only integers (samples with no fractional component) are added to the collection, requests for a quantile will round to the nearest integer.

**Exact values for sparse bins**. If only one value has been added to a certain bin, instead of the midpoint between the bin's lower and upper value being used, the exact sample value will be used. 

*Changed in version 0.3.0*: If repeated additions of the same value are made and all match this accurate sample value, the accurate sample will be retained. The first differing value to be added to that bin will cause the bin's midpoint value to be used instead.

## Categorizing the Quantogram Algorithm

Measuring quantiles of streams of data (such as the median) is an essential basic statistic in many applications. Your choices are:

  1. Accurate and fast with moderate storage requirements, limited to a dense range of values like integers
  2. Accurate and slow with high storage requirements
  3. Approximate with probabilistic guarantees of accuracy,  fast, and low storage requirements
  4. Approximate with fixed guarantees of accuracy, fast, and moderate storage requirements

**Dense histogram**. The first is ideal but has limited applicability. It is typically implemented using a list as a dense histogram with each bucket corresponding to an integer value. If all your samples are integers that fall beween zero and 10,000 (or are floats on a uniform grid with no more than 10,000 grid lines) it is the best choice, both easy to implement and efficient to use.

**Sorted list**. The second includes the naive approach, which is to gather all the data in a list, then sort it to get the quantile. To expedite, you can use **quickselect** so that you do not need to sort all the data. This approach does not work for streaming data, because the storage requirements are vast. 

**Probabilistic**. The third category has algorithms (such as variations on **Count Min Sketch**) that guarantee a given accuracy with a certain probability. A popular median estimator is called **Median of Medians**. If you know how many items N there will be, you can use **reservoir sampling** to maintain a sample set of the data and take the median of that. However, the more extreme your quantile, the larger a subset you must keep. With these algorithms, you usually get a good estimate, but there is always a chance that you get a bad estimate. Also, given the same set of data, you may get different results; the answer is not deterministic.


changelog.md
# Quantogram Release Notes


**Release 0.4.4**:

Fixed bug in quantile_at. When passed a NaN, it formerly panicked.

**Release 0.4.3**:

Speed up hsm (half-sample mode), cache its values for reuse and also expose it as a function that can be called even without a Quantogram. The cache makes the computation 9x faster than before once the number of samples exceeds 1000.

**Release 0.4.2**:

Remove debug logging statement from hsm.

**Release 0.4.1**:

Fix half-sample mode (hsm) edge case where there is no true mode.

**Release 0.4.0**:

Added these basic statistics:

 - **q1** (1st quartile)
 - **q3** (3rd quartile)

Added these measures of dispersion: 

 - **range** (spread between min and max)
 - **iqr** (interquartile range: Q3 - Q1)        
 - **quartile_deviation** (semi-interquartile range, half of iqr))
 - **coeff_of_range** ((max - min) / (max + min))
 - **coeff_of_quartile_dev** ((Q3-Q1)/(Q3+Q1))
 - **coeff_of_stddev** (standard deviation divided by mean)
 - **coeff_of_variation**

**Release 0.3.0**:

 - Added the **half-sample mode** measure, `Quantogram::hsm`, an estimate of the mode that is resistant to outliers. Based on the Robertson-Cryer half-sample mode.
 - If adding the same sample multiple times and a bin has only one distinct value that matches that accurate value, keep using the accurate value and do not switch to using the bin's midpoint value.

  
**Release 0.2.0**:

 - Added Quantogram::variance
 - Added Quantogram::stddev (standard deviation)
 - Quantogram::add_unweighted_samples now accepts iterators over any value that can convert to an f64 via the Into trait.

src/half_sample_mode.rs
use std::cmp::Ordering;
use super::agreement::Agreement;

// Contains: 
// 
//   half_sample_mode - Computes the half-sample mode, an estimate of the mode for weighted data.
//   HalfSampleInterval - internal struct
//   HalfSampleModeCache - Caches prior result of half_sample_mode to focus recalculation of mode on a subset of data, a time optimization. 

/// Internal structure for Half Sample Mode calculation.
#[derive(Copy, Clone, Debug)]
struct HalfSampleInterval {
    /// Index of the first bin in rthe interval
    pub start_index: usize,

    /// Sample value for first bin.
    pub start_sample: f64,

    /// Count of the bins in the interval
    pub bins: usize,

    /// Sum of weights from all bins in the interval
    pub cume_weight: f64,

    /// Width of interval is the difference between the lowest bin sample value 
    /// and the highest bin sample value in the interval.
    pub width: f64
}

impl HalfSampleInterval {
    pub fn new(index: usize, sample: f64, weight: f64) -> Self {
        HalfSampleInterval {
            start_index: index,
            start_sample: sample,
            bins: 1,
            cume_weight: weight,
            width: 0.0
        }
    }

    pub fn reset(&mut self, index: usize, sample: f64, weight: f64) {
        self.start_index = index;
        self.start_sample = sample;
        self.bins = 1;
        self.cume_weight = weight;
        self.width = 0.0;
    }

    pub fn empty() -> Self {
        Self::new(0, 0.0, 0.0)
    }

    pub fn is_empty(&self) -> bool {
        self.cume_weight == 0.0
    }

    pub fn add(&mut self, sample: f64, weight: f64) {
        self.bins += 1;
        self.cume_weight += weight;
        self.width = sample - self.start_sample;
    }

    pub fn is_underweight(&self, minimum_weight: f64) -> bool {
        self.cume_weight < minimum_weight
    }

    /// Check if this interval is denser than the other interval. 
    pub fn is_denser_than(&self, other: &Self, minimum_weight: f64) -> bool {
        if other.is_empty() && !self.is_empty() { true }
        else if self.is_underweight(minimum_weight) { false }
        else if other.is_underweight(minimum_weight) { true }
        else if self.width < other.width { true }
        else if self.width > other.width { false }
        else { self.cume_weight > other.cume_weight }
    }

}

/// Estimate the mode using the half-sample mode algorithm.
/// 
/// More resistant than the mode to outliers and contamination (noise). 
/// 
/// Based on the mode estimator by Robertson and Cryer (1974), and an 
/// algorithm described in "On a Fast, Robust Estimator of the Mode" by David Bickel and Rudolf Fruhwirth.
/// That algorithm is applied to raw samples, whereas this is applied 
/// to already histogrammed values that are weighted.
/// 
/// NOTE: This has worse than quadratic performance, possibly N^2 Log N.
/// 
/// # Arguments
/// 
/// * `samples` - List of sample values, sorted in ascending order.
/// * `weights` - List of positive weights, corresponding to samples.
/// * `total_weight` - Sum of the weights.
pub fn half_sample_mode(samples: &[f64], weights: &[f64], total_weight: f64) -> Option<f64> {
    if total_weight == 0.0 { None }
    else {
        let count = samples.len();
        if count < 5 {
            let mut max_weight = 0.0;
            let mut samples_with_max_weight: Vec<f64> = Vec::new();
            for (&sample, &weight) in samples.iter().zip(weights) {
                if weight > max_weight {
                    max_weight = weight;
                    samples_with_max_weight = vec![sample];
                }
                else if weight == max_weight {
                    samples_with_max_weight.push(sample);
                }
            }
            return Some(samples_with_max_weight.iter().sum::<f64>() / samples_with_max_weight.len() as f64);
        } 
        let mut cume_target = total_weight / 2.0;
        let mut start_index = 0;
        let mut limit_index = samples.len();
        let mut densest = HalfSampleInterval::empty();
        while (limit_index - start_index) >= 2 { 
            densest = HalfSampleInterval::empty();
            for interval_start in start_index..limit_index {
                let mut current_interval = HalfSampleInterval::empty();
                let mut made_weight = false;
                for index in interval_start..limit_index {
                    if current_interval.is_empty() {
                        current_interval.reset(index, samples[index], weights[index]);
                    }
                    else {
                        current_interval.add(samples[index], weights[index]);
                    }
                    if current_interval.is_denser_than(&densest, cume_target) {
                        densest = current_interval;
                    }
                    if !current_interval.is_underweight(cume_target) { 
                        made_weight = true;
                        break; 
                    }     
                }
                // Even going all the way to limit_index did not accumulate enough
                // weight to equal cume_target, so we can be done with the middle loop.
                if !made_weight {
                    break;
                }
            }
            start_index = densest.start_index;
            limit_index = start_index + densest.bins;
            cume_target = cume_target / 2.0;
        }
        match densest.bins {
            0 => None,
            1 => Some(samples[densest.start_index]),
            2 => {
                // TODO: Instead of taking the weighted average of two points, should we take the maximum?
                let i = densest.start_index;
                let weight_sum = weights[i] + weights[i+1];
                if weight_sum <= 0.0 {
                    None
                }
                else {
                    let weighted_mean = (samples[i] * weights[i] + samples[i+1] * weights[i+1]) / weight_sum;
                    Some(weighted_mean)
                }
            },
            3 => {
                // TODO: Instead of taking the weighted average of three points, should we take the maximum?
                let i = densest.start_index;
                let weight_sum = weights[i] + weights[i+1] + weights[i+2];
                if weight_sum <= 0.0 {
                    None
                }
                else {
                    let weighted_mean = (
                        samples[i] * weights[i] 
                      + samples[i+1] * weights[i+1]
                      + samples[i+2] * weights[i+2]) 
                      / weight_sum;
                    Some(weighted_mean)
                }
            },
            _ => None
        }
    }
}

/// Permit the caching and reuse of mode values computed using half-sample mode
/// in order to enhance performance. 
/// 
/// half_sample_mode (hsm) is a costly function, worse than quadratic.
/// 
/// Cache the last computed value of the half-sample mode and 
/// use it to restrict the range of samples to be reazalyzed
/// when you compute a new mode after one or more samples have been inserted. 
/// 
/// If the new sample is in the same neighborhood as the previous mode, 
/// assume that the new mode must be in the same neighborhood and
/// only make computations on samples in that neighborhood.
/// 
/// The cache is invalidated if: 
///   - the new sample is out of the neighborhood
///   - the total_weight has changed too much
///   - the number of bins has changed too much
#[derive(Copy, Clone, Debug)]
pub struct HalfSampleModeCache {
    /// Previous mode computed by hsm, or none if invalidated.
    mode: Option<f64>,

    /// Most recent sample added to date.
    latest_sample: f64,

    /// Alternate mode estimate (thst is faster to compute)
    latest_proxy: f64,

    /// Number of bins present during previous full recalculation of mode.
    bins: usize,

    /// Total weight of all samples during previous full recalculation of mode.
    total_weight: f64,

    /// Number of consecutive bins considered near to the current mode, half below and half above.
    neighborhood: usize,

    /// If the number of bins has changed by more than this many, invalidate the cache.
    allowed_bin_change: usize,

    /// If the relative change in total_weight exceeds this, invalidate the cache.
    allowed_weight_change_fraction: f64,

    allowed_outside_neighborhood: usize,

    outside_neighborhood_counter: usize,

    cache_hits: usize,

    /// Measures the agreement of the mode and hsm.
    /// Currently not useful, as those two measures are not in good enough agreement (based on experiment)
    /// to permit mode to be used in place of hsm. The goal is to find a less expensive estimate of mode (a proxy) than
    /// hsm, track its agreement, and if the agreement after 1,000 iterations is good, to 
    /// swap in use of the proxy to improve performance.
    agreement: Agreement
}

impl HalfSampleModeCache {
    pub fn new(neighborhood: usize, allowed_bin_change: usize, allowed_weight_change_fraction: f64) -> Self {
        HalfSampleModeCache {
            mode: None,
            latest_sample: f64::NAN,
            latest_proxy: f64::NAN,
            bins: 0,
            total_weight: 0.0,
            neighborhood: neighborhood,
            allowed_bin_change: allowed_bin_change,
            allowed_weight_change_fraction: allowed_weight_change_fraction,
            allowed_outside_neighborhood: 12,
            outside_neighborhood_counter: 0,
            cache_hits: 0,
            agreement: Agreement::new()
        }
    }

    pub fn new_default() -> Self {
        HalfSampleModeCache {
            mode: None,
            latest_sample: f64::NAN,
            latest_proxy: f64::NAN,
            bins: 0,
            total_weight: 0.0,
            // A neighborhood of 70 is chosen so that if a growth factor of 1.02 is used
            // in a Quantogram for 1% error, we will range over data in the 
            // range [1/2 x mode, 2 x mode] because there are then 35 bins per doubling,
            // which should capture sufficient range.
            neighborhood: 70,
            allowed_bin_change: 10,
            allowed_weight_change_fraction: 0.01,
            allowed_outside_neighborhood: 12,
            outside_neighborhood_counter: 0,
            cache_hits: 0,
            agreement: Agreement::new()
        }
    }

    /// Record the latest sample added to data.
    /// This must be called prior to calling hsm so that we know if the new value is near the prior mode.
    pub fn record(&mut self, sample: f64, mode_proxy: f64) {
        self.latest_sample = sample;
        self.latest_proxy = mode_proxy;
    }

    pub fn hits(&self) -> usize { self.cache_hits }

    /// Assumes that latest_proxy was set by a call to record(), 
    /// then mode was set in one of the recalc methods.
    fn update_agreement(&mut self) {
        if self.latest_proxy.is_finite() && self.mode.is_some() && self.latest_proxy.is_finite() {
            self.agreement.add(self.mode.unwrap(), self.latest_proxy);
            // Prevent reuse of the same proxy value; it will skew the results.
            self.latest_proxy = f64::NAN;
        }
    }

    fn full_recalc(&mut self, samples: &[f64], weights: &[f64], total_weight: f64) -> Option<f64> {
        self.bins = samples.len();
        self.total_weight = total_weight;
        self.mode = half_sample_mode(samples, weights, total_weight);
        self.outside_neighborhood_counter = 0;

        self.update_agreement();

        self.mode
    }

    /// Cribbed from the nightly feature on f64 to order floats.
    /// TODO: Once that feature becomes stable, should switch to using it.
    /// See #[unstable(feature = "total_cmp", issue = "72599")]
    #[inline]
    pub fn total_cmp(this: &f64, other: &f64) -> Ordering {
        let mut left = this.to_bits() as i64;
        let mut right = other.to_bits() as i64;

        // In case of negatives, flip all the bits except the sign
        // to achieve a similar layout as two's complement integers
        //
        // Why does this work? IEEE 754 floats consist of three fields:
        // Sign bit, exponent and mantissa. The set of exponent and mantissa
        // fields as a whole have the property that their bitwise order is
        // equal to the numeric magnitude where the magnitude is defined.
        // The magnitude is not normally defined on NaN values, but
        // IEEE 754 totalOrder defines the NaN values also to follow the
        // bitwise order. This leads to order explained in the doc comment.
        // However, the representation of magnitude is the same for negative
        // and positive numbers – only the sign bit is different.
        // To easily compare the floats as signed integers, we need to
        // flip the exponent and mantissa bits in case of negative numbers.
        // We effectively convert the numbers to "two's complement" form.
        //
        // To do the flipping, we construct a mask and XOR against it.
        // We branchlessly calculate an "all-ones except for the sign bit"
        // mask from negative-signed values: right shifting sign-extends
        // the integer, so we "fill" the mask with sign bits, and then
        // convert to unsigned to push one more zero bit.
        // On positive values, the mask is all zeros, so it's a no-op.
        left ^= (((left >> 63) as u64) >> 1) as i64;
        right ^= (((right >> 63) as u64) >> 1) as i64;

        left.cmp(&right)
    }

    /// Recalculate the half-sample mode but only examine a neighborhood around the previous
    /// estimate for mode, not the whole data set.
    fn partial_recalc(&mut self, prev_mode_position: usize, samples: &[f64], weights: &[f64], total_weight: f64) -> Option<f64> {
        self.cache_hits += 1;
        self.bins = samples.len();
        self.total_weight = total_weight;
        let half_neighborhood = self.neighborhood / 2;

        let start_index = if prev_mode_position < half_neighborhood { 0 }
        else { prev_mode_position - half_neighborhood };

        let stop_index = if prev_mode_position + half_neighborhood >= samples.len() { samples.len() }
        else { prev_mode_position + half_neighborhood };

        self.mode = half_sample_mode(&samples[start_index..stop_index], &weights[start_index..stop_index], total_weight);

        self.update_agreement();

        self.mode
    }

    /// Compute the half-sample mode of weighted data. 
    /// 
    /// A full recalculation or a partial recalculation will be performed, depending on
    /// whether the potential for the mode to have drifted far seems likely or not.
    /// 
    /// Before invoking this, you must call record(sample) to set the latest sample value.
    /// This is so that you can add several samples in between calls to hsm if necessary, then call hsm.
    /// The adder of samples should not need to know when or if hsm is to be called soon.
    pub fn hsm(&mut self, samples: &[f64], weights: &[f64], total_weight: f64) -> Option<f64> {
        let new_bins = samples.len();
        let do_full_recalc = self.mode.is_none()
          || new_bins < 64 // For fewer samples, no need to cut corners because fast enough
          || new_bins - self.bins > self.allowed_bin_change
          || (total_weight - self.total_weight) / self.total_weight > self.allowed_weight_change_fraction
        ;
        let prev_mode = self.mode.unwrap_or_default();
        if !do_full_recalc && prev_mode.is_finite() {
            let prev_mode_position = match samples.binary_search_by(|probe| Self::total_cmp(probe, &prev_mode)) {
                Result::Ok(pos) => pos,
                Err(pos) => pos
            };
            let newest_sample_position = match samples.binary_search_by(|probe| Self::total_cmp(probe, &self.latest_sample)) {
                Result::Ok(pos) => pos,
                Err(pos) => pos
            };
            // If the added sample is too far from the center of the neighborhood, we need a full recalc.
            let spread = usize::max(prev_mode_position, newest_sample_position) - usize::min(prev_mode_position, newest_sample_position);
            let third_neighborhood = self.neighborhood / 3;
            let half_neighborhood = self.neighborhood / 2;
            if spread > half_neighborhood && self.outside_neighborhood_counter < self.allowed_outside_neighborhood {
                // If the latest sample is outside the neighborhood, is unlikely to move the mode. 
                // Increment a counter so that we are forced to invalidate the cache 
                // if too many such samples are received. 
                self.outside_neighborhood_counter += 1;
                // Reuse previous value with no recalcuation.
                self.cache_hits += 1;
                self.mode
            }
            else if spread > third_neighborhood {
                if self.outside_neighborhood_counter >= self.allowed_outside_neighborhood {             
                    self.full_recalc(samples, weights, total_weight)
                }
                else {
                    self.outside_neighborhood_counter += 1;
                    self.partial_recalc(prev_mode_position, samples, weights, total_weight)
                }
            }
            else {
                self.partial_recalc(prev_mode_position, samples, weights, total_weight)
            }
        }
        else {
            self.full_recalc(samples, weights, total_weight)
        }
    }

    /// Clear the cache, forcing the next call to hsm to do a full recalc.
    pub fn clear(&mut self) {
        self.mode = None;
        self.latest_sample = f64::NAN;
        self.bins = 0;
        self.total_weight = f64::NAN;
        self.cache_hits = 0;
        self.outside_neighborhood_counter = 0;
    }
}


#[cfg(test)]
mod tests {
    use std::time::{Duration, Instant};
    use crate::{Quantogram, QuantogramBuilder};
    use super::*;

    #[test]
    fn hsm_cache_performance() {
        let no_cache_duration = simulate_convergence(1200, false).as_secs_f64();
        let with_cache_duration = simulate_convergence(1200, true).as_secs_f64();
        let speedup = no_cache_duration / with_cache_duration;
        println!("
        With cache:    {:.2} sec
        Without Cache: {:.2} sec
        Speedup:       {:.3}x
        ", with_cache_duration, no_cache_duration, speedup);
        assert!(speedup > 8.0);
    }

    /// Golden ratio based pseudo-random number sequence.
    ///  ```text
    ///    φ = (1 + √5) / 2
    ///    G(n) = (n · φ) mod 1
    /// ```
    fn golden_ratio_sequence(n: usize) -> f64 {
        let phi = (1.0 + 5.0_f64.sqrt()) / 2.0;
        (n as f64 * phi) % 1.0
    }

    fn insert_sorted(value: f64, samples: &mut Vec<f64>) {
        let pos = match samples.binary_search_by(|probe| HalfSampleModeCache::total_cmp(probe, &value)) {
            Result::Ok(pos) => pos,
            Err(pos) => pos
        };
        samples.insert(pos, value);
    }

    /// Simulate a series of estimates of a value that steadily converge. 
    /// That means that a modal peak will grow near the true solution.
    fn simulate_convergence(iterations: usize, use_cache: bool) -> Duration {
        let true_value = 100.0;
        let start_value = 0.0;
        let start_error = start_value - true_value;
        let mut values: Vec<f64> = Vec::new();
        let mut weights: Vec<f64> = Vec::new();
        let mut weight_sum = 0.0;
        let mut cache = HalfSampleModeCache::new_default();
        let start_time = Instant::now();
        let mut q = QuantogramBuilder::new().with_error(0.01).build();
        for i in 1..=iterations {
            let error = start_error / (i as f64).sqrt();
            let sign = if golden_ratio_sequence(i) < 0.3 { -1.0 } else { 1.0 };
            let next_value = true_value + sign * error;
            // Must ensure values remain sorted.
            insert_sorted(next_value, &mut values);
            q.add(next_value);
            // values.push(next_value);
            weights.push(1.0);
            weight_sum += 1.0;
            if use_cache {
                cache.record(next_value, next_value);
                let _mode = cache.hsm(&values, &weights, weight_sum);
            }
            else {
                let _mode = half_sample_mode(&values, &weights, weight_sum);
            }
        }
        if use_cache {
            println!("  Cache hits: {}. Cache misses: {}", cache.hits(), iterations - cache.hits());
        }
        start_time.elapsed()
    }

}